# Pandas and Covid-19

This notebook is an example of data analysis and manipulation with **Pandas** and has beed created in [Google Colab](https://colab.research.google.com)

Enjoy it!

In [ ]:
import numpy as np
import pandas as pd

## Data

To get some data I and going to download it from Data Repository by Johns Hopkins CSSE

https://github.com/CSSEGISandData/COVID-19

I like to remove the folder where I am going to save the data so I can re-execute this sentences without any problems ...

In [ ]:
!rm -rf ./COVID-19

The dataset is avaible in GitHub so I use the `git` command to get it

In [ ]:
!git clone https://github.com/CSSEGISandData/COVID-19.git

Cloning into 'COVID-19'...
remote: Enumerating objects: 390558, done.
remote: Counting objects: 100% (106/106), done.
remote: Compressing objects: 100% (54/54), done.
remote: Total 390558 (delta 67), reused 86 (delta 52), pack-reused 390452
Receiving objects: 100% (390558/390558), 4.78 GiB | 27.84 MiB/s, done.
Resolving deltas: 100% (324871/324871), done.
Checking out files: 100% (1876/1876), done.


More than **4.78 Gb** of data ...

## Exporing the data


### Exploring the files

The first step is explore the datafiles. I will use command line orders to save the filenames into a file

In [ ]:
!ls -lt ./COVID-19/csse_covid_19_data/csse_covid_19_daily_reports  | head

total 423240
-rw-r--r-- 1 root root      0 Jun  4 08:38 README.md
-rw-r--r-- 1 root root 558819 Jun  4 08:38 12-31-2021.csv
-rw-r--r-- 1 root root 569606 Jun  4 08:38 12-31-2020.csv
-rw-r--r-- 1 root root 558746 Jun  4 08:38 12-30-2021.csv
-rw-r--r-- 1 root root 569588 Jun  4 08:38 12-30-2020.csv
-rw-r--r-- 1 root root 558742 Jun  4 08:38 12-29-2021.csv
-rw-r--r-- 1 root root 569580 Jun  4 08:38 12-29-2020.csv
-rw-r--r-- 1 root root 558738 Jun  4 08:38 12-28-2021.csv
-rw-r--r-- 1 root root 569414 Jun  4 08:38 12-28-2020.csv


In [ ]:
! ls -l ./COVID-19/csse_covid_19_data/csse_covid_19_daily_reports > files.txt

In [21]:
! cat files.txt | head

total 423240
-rw-r--r-- 1 root root 569597 Jun  4 08:38 01-01-2021.csv
-rw-r--r-- 1 root root 558819 Jun  4 08:38 01-01-2022.csv
-rw-r--r-- 1 root root 569827 Jun  4 08:38 01-02-2021.csv
-rw-r--r-- 1 root root 558641 Jun  4 08:38 01-02-2022.csv
-rw-r--r-- 1 root root 570000 Jun  4 08:38 01-03-2021.csv
-rw-r--r-- 1 root root 558850 Jun  4 08:38 01-03-2022.csv
-rw-r--r-- 1 root root 570011 Jun  4 08:38 01-04-2021.csv
-rw-r--r-- 1 root root 558937 Jun  4 08:38 01-04-2022.csv
-rw-r--r-- 1 root root 570185 Jun  4 08:38 01-05-2021.csv


I will load this file into a Pandas dataframe and clean it

In [58]:
df_data_files = pd.read_fwf("files.txt", header = None)

In [46]:
df_data_files.head(10)

0     1     2         3    4    5      6               7
0  total 423240   NaN   NaN       NaN  NaN  NaN    NaN             NaN
1  -rw-r--r-- 1  root  root  569597.0  Jun  4.0  08:38  01-01-2021.csv
2  -rw-r--r-- 1  root  root  558819.0  Jun  4.0  08:38  01-01-2022.csv
3  -rw-r--r-- 1  root  root  569827.0  Jun  4.0  08:38  01-02-2021.csv
4  -rw-r--r-- 1  root  root  558641.0  Jun  4.0  08:38  01-02-2022.csv
5  -rw-r--r-- 1  root  root  570000.0  Jun  4.0  08:38  01-03-2021.csv
6  -rw-r--r-- 1  root  root  558850.0  Jun  4.0  08:38  01-03-2022.csv
7  -rw-r--r-- 1  root  root  570011.0  Jun  4.0  08:38  01-04-2021.csv
8  -rw-r--r-- 1  root  root  558937.0  Jun  4.0  08:38  01-04-2022.csv
9  -rw-r--r-- 1  root  root  570185.0  Jun  4.0  08:38  01-05-2021.csv

In [69]:
df_files = df_data_files.rename(columns= { 7: 'filename'}) \
  .filter(['filename']) \
  .query("filename != 'README.md' and filename.notnull()", engine = 'python') \
  .assign(date = lambda dataset: pd.to_datetime(dataset.filename.str[0:10], format = "%m-%d-%Y"),
          month = lambda dataset: dataset.date.dt.month,
          year = lambda dataset: dataset.date.dt.year)

In [74]:
df_files.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 864 entries, 1 to 864
Data columns (total 4 columns):
 #   Column    Non-Null Count  Dtype         
---  ------    --------------  -----         
 0   filename  864 non-null    object        
 1   date      864 non-null    datetime64[ns]
 2   month     864 non-null    int64         
 3   year      864 non-null    int64         
dtypes: datetime64[ns](1), int64(2), object(1)
memory usage: 33.8+ KB


In [70]:
df_files

filename       date  month  year
1    01-01-2021.csv 2021-01-01      1  2021
2    01-01-2022.csv 2022-01-01      1  2022
3    01-02-2021.csv 2021-01-02      1  2021
4    01-02-2022.csv 2022-01-02      1  2022
5    01-03-2021.csv 2021-01-03      1  2021
..              ...        ...    ...   ...
860  12-29-2021.csv 2021-12-29     12  2021
861  12-30-2020.csv 2020-12-30     12  2020
862  12-30-2021.csv 2021-12-30     12  2021
863  12-31-2020.csv 2020-12-31     12  2020
864  12-31-2021.csv 2021-12-31     12  2021

[864 rows x 4 columns]

In [71]:
df_files.groupby("year").agg(total_files = ("filename", "count"))

total_files
year             
2020          345
2021          365
2022          154

Now I can know the first and the last datafile

In [75]:
df_files \
  .sort_values(['date']) \
  .head(5)

filename       date  month  year
43  01-22-2020.csv 2020-01-22      1  2020
46  01-23-2020.csv 2020-01-23      1  2020
49  01-24-2020.csv 2020-01-24      1  2020
52  01-25-2020.csv 2020-01-25      1  2020
55  01-26-2020.csv 2020-01-26      1  2020

In [76]:
df_files \
  .sort_values(['date'], ascending = False) \
  .head(5)

filename       date  month  year
442  06-03-2022.csv 2022-06-03      6  2022
439  06-02-2022.csv 2022-06-02      6  2022
436  06-01-2022.csv 2022-06-01      6  2022
433  05-31-2022.csv 2022-05-31      5  2022
430  05-30-2022.csv 2022-05-30      5  2022

### Exploring the data

Perfect. Let's explore the first dataset generated ...

In [77]:
first = pd.read_csv("./COVID-19/csse_covid_19_data/csse_covid_19_daily_reports/01-22-2020.csv")

In [78]:
first.head()

Province/State  Country/Region      Last Update  Confirmed  Deaths  \
0          Anhui  Mainland China  1/22/2020 17:00        1.0     NaN   
1        Beijing  Mainland China  1/22/2020 17:00       14.0     NaN   
2      Chongqing  Mainland China  1/22/2020 17:00        6.0     NaN   
3         Fujian  Mainland China  1/22/2020 17:00        1.0     NaN   
4          Gansu  Mainland China  1/22/2020 17:00        NaN     NaN   

   Recovered  
0        NaN  
1        NaN  
2        NaN  
3        NaN  
4        NaN

In [80]:
first.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 66 entries, 0 to 65
Data columns (total 6 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Province/State  56 non-null     object 
 1   Country/Region  66 non-null     object 
 2   Last Update     66 non-null     object 
 3   Confirmed       56 non-null     float64
 4   Deaths          29 non-null     float64
 5   Recovered       29 non-null     float64
dtypes: float64(3), object(3)
memory usage: 3.2+ KB


And the last one ...

In [81]:
last = pd.read_csv("./COVID-19/csse_covid_19_data/csse_covid_19_daily_reports/06-03-2022.csv")

In [82]:
last.head()

FIPS Admin2 Province_State Country_Region          Last_Update       Lat  \
0   NaN    NaN            NaN    Afghanistan  2022-06-04 04:20:58  33.93911   
1   NaN    NaN            NaN        Albania  2022-06-04 04:20:58  41.15330   
2   NaN    NaN            NaN        Algeria  2022-06-04 04:20:58  28.03390   
3   NaN    NaN            NaN        Andorra  2022-06-04 04:20:58  42.50630   
4   NaN    NaN            NaN         Angola  2022-06-04 04:20:58 -11.20270   

       Long_  Confirmed  Deaths  Recovered  Active Combined_Key  \
0  67.709953     180584    7708        NaN     NaN  Afghanistan   
1  20.168300     276310    3497        NaN     NaN      Albania   
2   1.659600     265889    6875        NaN     NaN      Algeria   
3   1.521800      43067     153        NaN     NaN      Andorra   
4  17.873900      99761    1900        NaN     NaN       Angola   

   Incident_Rate  Case_Fatality_Ratio  
0     463.888250             4.268374  
1    9601.431649             1.265607  
2     606.345897             2.585665  
3   55739.338640             0.355260  
4     303.536136             1.904552

In [83]:
last.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4011 entries, 0 to 4010
Data columns (total 14 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   FIPS                 3268 non-null   float64
 1   Admin2               3272 non-null   object 
 2   Province_State       3834 non-null   object 
 3   Country_Region       4011 non-null   object 
 4   Last_Update          4011 non-null   object 
 5   Lat                  3921 non-null   float64
 6   Long_                3921 non-null   float64
 7   Confirmed            4011 non-null   int64  
 8   Deaths               4011 non-null   int64  
 9   Recovered            0 non-null      float64
 10  Active               0 non-null      float64
 11  Combined_Key         4011 non-null   object 
 12  Incident_Rate        3918 non-null   float64
 13  Case_Fatality_Ratio  3971 non-null   float64
dtypes: float64(7), int64(2), object(5)
memory usage: 438.8+ KB


Can I concatenate both datasets?

In [87]:
pd.concat((first, last))

Province/State  Country/Region      Last Update  Confirmed  Deaths  \
0             Anhui  Mainland China  1/22/2020 17:00        1.0     NaN   
1           Beijing  Mainland China  1/22/2020 17:00       14.0     NaN   
2         Chongqing  Mainland China  1/22/2020 17:00        6.0     NaN   
3            Fujian  Mainland China  1/22/2020 17:00        1.0     NaN   
4             Gansu  Mainland China  1/22/2020 17:00        NaN     NaN   
...             ...             ...              ...        ...     ...   
4006            NaN             NaN              NaN   657705.0  5660.0   
4007            NaN             NaN              NaN      535.0     0.0   
4008            NaN             NaN              NaN    11822.0  2149.0   
4009            NaN             NaN              NaN   322207.0  3988.0   
4010            NaN             NaN              NaN   253236.0  5509.0   

      Recovered  FIPS Admin2 Province_State        Country_Region  \
0           NaN   NaN    NaN            NaN                   NaN   
1           NaN   NaN    NaN            NaN                   NaN   
2           NaN   NaN    NaN            NaN                   NaN   
3           NaN   NaN    NaN            NaN                   NaN   
4           NaN   NaN    NaN            NaN                   NaN   
...         ...   ...    ...            ...                   ...   
4006        NaN   NaN    NaN            NaN    West Bank and Gaza   
4007        NaN   NaN    NaN            NaN  Winter Olympics 2022   
4008        NaN   NaN    NaN            NaN                 Yemen   
4009        NaN   NaN    NaN            NaN                Zambia   
4010        NaN   NaN    NaN            NaN              Zimbabwe   

              Last_Update        Lat       Long_  Active  \
0                     NaN        NaN         NaN     NaN   
1                     NaN        NaN         NaN     NaN   
2                     NaN        NaN         NaN     NaN   
3                     NaN        NaN         NaN     NaN   
4                     NaN        NaN         NaN     NaN   
...                   ...        ...         ...     ...   
4006  2022-06-04 04:20:58  31.952200   35.233200     NaN   
4007  2022-06-04 04:20:58  39.904200  116.407400     NaN   
4008  2022-06-04 04:20:58  15.552727   48.516388     NaN   
4009  2022-06-04 04:20:58 -13.133897   27.849332     NaN   
4010  2022-06-04 04:20:58 -19.015438   29.154857     NaN   

              Combined_Key  Incident_Rate  Case_Fatality_Ratio  
0                      NaN            NaN                  NaN  
1                      NaN            NaN                  NaN  
2                      NaN            NaN                  NaN  
3                      NaN            NaN                  NaN  
4                      NaN            NaN                  NaN  
...                    ...            ...                  ...  
4006    West Bank and Gaza   12892.596879             0.860568  
4007  Winter Olympics 2022            NaN             0.000000  
4008                 Yemen      39.636601            18.177973  
4009                Zambia    1752.653237             1.237714  
4010              Zimbabwe    1703.809754             2.175441  

[4077 rows x 17 columns]

Ups!!! The column names don't match :-(

## Loading the data into Pandas and cleaning it

In [88]:
import glob
import os

files = glob.glob("./COVID-19/csse_covid_19_data/csse_covid_19_daily_reports/*.csv")
files.sort(key=os.path.getmtime)

We are going to:
- Create a blank Dataset to store all the data
- Load every dataset unifying the column names so we can concatenate it without any problem.
- Remove extra blank spaces from the country field
- Enrich the information with the date of the data in the correct type

In [159]:
df_all_data = pd.DataFrame()
for file in files:  
    df_file = pd.read_csv(file).rename(columns = {'Province/State' : 'State', 
                        'Province_State' : 'State', 
                        "Country/Region" : 'Country',
                        "Country_Region" : 'Country',
                        'Last Update' : 'Last_Update',
                        'Confirmed' : 'ConfirmedAcum',
                        'Deaths' : 'DeathsAcum',
                        'Recovered' : 'RecoveredAcum'})
    df_file = df_file.assign(Date = pd.to_datetime(file[-14:-4], format = '%m-%d-%Y'),
                 Country = lambda dataset: dataset.Country.str.strip())
    df_all_data = pd.concat([df_all_data, df_file])


Let's create a new variable in case it is necessary to repeat the analysis.

In [180]:
df_data = df_all_data.copy()

In [181]:
df_data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 3162110 entries, 0 to 4010
Data columns (total 19 columns):
 #   Column               Dtype         
---  ------               -----         
 0   FIPS                 float64       
 1   Admin2               object        
 2   State                object        
 3   Country              object        
 4   Last_Update          object        
 5   Lat                  float64       
 6   Long_                float64       
 7   ConfirmedAcum        float64       
 8   DeathsAcum           float64       
 9   RecoveredAcum        float64       
 10  Active               float64       
 11  Combined_Key         object        
 12  Incident_Rate        float64       
 13  Case_Fatality_Ratio  float64       
 14  Date                 datetime64[ns]
 15  Latitude             float64       
 16  Longitude            float64       
 17  Incidence_Rate       float64       
 18  Case-Fatality_Ratio  float64       
dtypes: datetime64[ns](1), fl

In [182]:
df_data.head(10)

FIPS Admin2                         State              Country  \
0   NaN    NaN                           NaN          Afghanistan   
1   NaN    NaN                           NaN              Albania   
2   NaN    NaN                           NaN              Algeria   
3   NaN    NaN                           NaN              Andorra   
4   NaN    NaN                           NaN               Angola   
5   NaN    NaN                           NaN  Antigua and Barbuda   
6   NaN    NaN                           NaN            Argentina   
7   NaN    NaN                           NaN              Armenia   
8   NaN    NaN  Australian Capital Territory            Australia   
9   NaN    NaN               New South Wales            Australia   

           Last_Update       Lat       Long_  ConfirmedAcum  DeathsAcum  \
0  2021-01-02 05:22:33  33.93911   67.709953        52513.0      2201.0   
1  2021-01-02 05:22:33  41.15330   20.168300        58316.0      1181.0   
2  2021-01-02 05:22:33  28.03390    1.659600        99897.0      2762.0   
3  2021-01-02 05:22:33  42.50630    1.521800         8117.0        84.0   
4  2021-01-02 05:22:33 -11.20270   17.873900        17568.0       405.0   
5  2021-01-02 05:22:33  17.06080  -61.796400          159.0         5.0   
6  2021-01-02 05:22:33 -38.41610  -63.616700      1629594.0     43319.0   
7  2021-01-02 05:22:33  40.06910   45.038200       159738.0      2828.0   
8  2021-01-02 05:22:33 -35.47350  149.012400          118.0         3.0   
9  2021-01-02 05:22:33 -33.86880  151.209300         4947.0        54.0   

   RecoveredAcum    Active                             Combined_Key  \
0        41727.0    8585.0                              Afghanistan   
1        33634.0   23501.0                                  Albania   
2        67395.0   29740.0                                  Algeria   
3         7463.0     570.0                                  Andorra   
4        11146.0    6017.0                                   Angola   
5          148.0       6.0                      Antigua and Barbuda   
6      1426676.0  159599.0                                Argentina   
7       143355.0   13555.0                                  Armenia   
8          114.0       1.0  Australian Capital Territory, Australia   
9            0.0    4893.0               New South Wales, Australia   

   Incident_Rate  Case_Fatality_Ratio       Date  Latitude  Longitude  \
0     134.896578             4.191343 2021-01-01       NaN        NaN   
1    2026.409062             2.025173 2021-01-01       NaN        NaN   
2     227.809861             2.764848 2021-01-01       NaN        NaN   
3   10505.403482             1.034865 2021-01-01       NaN        NaN   
4      53.452981             2.305328 2021-01-01       NaN        NaN   
5     162.364186             3.144654 2021-01-01       NaN        NaN   
6    3605.633332             2.658269 2021-01-01       NaN        NaN   
7    5390.664389             1.770399 2021-01-01       NaN        NaN   
8      27.563653             2.542373 2021-01-01       NaN        NaN   
9      60.938655             1.091571 2021-01-01       NaN        NaN   

   Incidence_Rate  Case-Fatality_Ratio  
0             NaN                  NaN  
1             NaN                  NaN  
2             NaN                  NaN  
3             NaN                  NaN  
4             NaN                  NaN  
5             NaN                  NaN  
6             NaN                  NaN  
7             NaN                  NaN  
8             NaN                  NaN  
9             NaN                  NaN

First, I am going to filter the columns that are interesting to me

In [183]:
df_data = df_data.filter(['Date', 'Country', 'ConfirmedAcum', 'DeathsAcum', 'RecoveredAcum'])

In [184]:
df_data

Date               Country  ConfirmedAcum  DeathsAcum  \
0    2021-01-01           Afghanistan        52513.0      2201.0   
1    2021-01-01               Albania        58316.0      1181.0   
2    2021-01-01               Algeria        99897.0      2762.0   
3    2021-01-01               Andorra         8117.0        84.0   
4    2021-01-01                Angola        17568.0       405.0   
...         ...                   ...            ...         ...   
4006 2021-12-31  Winter Olympics 2022            0.0         0.0   
4007 2021-12-31            Antarctica           11.0         0.0   
4008 2021-12-31        United Kingdom        19002.0        89.0   
4009 2021-12-31        United Kingdom         7802.0        24.0   
4010 2021-12-31          Korea, North            0.0         0.0   

      RecoveredAcum  
0           41727.0  
1           33634.0  
2           67395.0  
3            7463.0  
4           11146.0  
...             ...  
4006            0.0  
4007            0.0  
4008            0.0  
4009            0.0  
4010            0.0  

[3162110 rows x 5 columns]

I realized that there were countries that were called by different names. 

Let's fix it ...

In [185]:
df_data = df_data.assign(
    Country = lambda dataset: dataset.Country.replace({'Bahamas, The' : 'Bahamas',
                         'Congo (Brazzaville)' : 'Congo',
                         'Congo (Kinshasa)' : 'Congo',
                         "Cote d'Ivoire" : "Cote d'Ivoire",
                         "Curacao" : "Curaçao",
                         'Czech Republic' : 'Czech Republic (Czechia)',
                         'Czechia' : 'Czech Republic (Czechia)',
                         'Faroe Islands' : 'Faeroe Islands',
                         'Macau' : 'Macao',
                         'Mainland China' : 'China',
                         'Palestine' : 'State of Palestine',
                         'Reunion' : 'Réunion',
                         'Saint Kitts and Nevis' : 'Saint Kitts & Nevis',
                         'Sao Tome and Principe' : 'Sao Tome & Principe',
                         'US' : 'United States',
                         'Gambia, The' : 'Gambia',
                         'Hong Kong SAR' : 'Hong Kong',
                         'Korea, South' : 'South Korea',
                         'Macao SAR' : 'Macao',
                         'Taiwan*' : 'Taiwan',
                         'Viet Nam' : 'Vietnam',
                         'West Bank and Gaza' : 'State of Palestine',
                         'occupied Palestinian territory' : 'State of Palestine'
                         })
)    

At some point the data of the recovered persons was not published anymore, so I am going to keep only the rows that contain this data

In [186]:
df_data = df_data.query("Date <= '2021-08-04'")

Let's verify the structure of the dateset ...

In [187]:
df_data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1946881 entries, 0 to 4005
Data columns (total 5 columns):
 #   Column         Dtype         
---  ------         -----         
 0   Date           datetime64[ns]
 1   Country        object        
 2   ConfirmedAcum  float64       
 3   DeathsAcum     float64       
 4   RecoveredAcum  float64       
dtypes: datetime64[ns](1), float64(3), object(1)
memory usage: 89.1+ MB


More than **1.9 MM** of rows!

In [188]:
df_data.query("Country == 'Spain'") \
    .sort_values('Date', ascending = False) \
    .head()

Date Country  ConfirmedAcum  DeathsAcum  RecoveredAcum
600 2021-08-04   Spain       149510.0      3601.0         3772.0
604 2021-08-04   Spain        82694.0       842.0         1537.0
599 2021-08-04   Spain       724716.0     10254.0        10671.0
612 2021-08-04   Spain        36563.0       797.0         3107.0
609 2021-08-04   Spain         6400.0       117.0          163.0

Wait a sec, I think that can be interesting have a column the the active cases. Let's create it ...

In [189]:
df_data = df_data.assign(
    ActiveAcum = lambda dataset: dataset.ConfirmedAcum  - dataset.DeathsAcum - dataset.RecoveredAcum
)

In [190]:
df_data.query("Country == 'Spain'") \
    .sort_values('Date', ascending = False) \
    .head()

Date Country  ConfirmedAcum  DeathsAcum  RecoveredAcum  ActiveAcum
600 2021-08-04   Spain       149510.0      3601.0         3772.0    142137.0
604 2021-08-04   Spain        82694.0       842.0         1537.0     80315.0
599 2021-08-04   Spain       724716.0     10254.0        10671.0    703791.0
612 2021-08-04   Spain        36563.0       797.0         3107.0     32659.0
609 2021-08-04   Spain         6400.0       117.0          163.0      6120.0

Perfect :-)

Now, I am going to group and summarize the data because I want to be sure that there is only one row per Date and Country

In [191]:
df_data = df_data \
      .groupby(["Date", "Country"], as_index = False )\
      .agg("sum")

In [192]:
df_data.query("Country == 'Spain'") \
    .sort_values('Date', ascending = False) \
    .head()

Date Country  ConfirmedAcum  DeathsAcum  RecoveredAcum  \
102187 2021-08-04   Spain      4545184.0     81844.0       150376.0   
101989 2021-08-03   Spain      4523310.0     81773.0       150376.0   
101791 2021-08-02   Spain      4502983.0     81643.0       150376.0   
101593 2021-08-01   Spain      4447044.0     81486.0       150376.0   
101395 2021-07-31   Spain      4447044.0     81486.0       150376.0   

        ActiveAcum  
102187   4312964.0  
101989   4291161.0  
101791   4270964.0  
101593   4215182.0  
101395   4215182.0

##  Daily Cases
I am going to enrich the data by creating new columns with the daily cases.  

In [193]:
df_data \
      .query("Country == 'Spain'") \
      .sort_values(['Country', 'Date'], ascending = False) \
      .filter(['Date', 'Country','ConfirmedAcum']) \
      .head(5)

Date Country  ConfirmedAcum
102187 2021-08-04   Spain      4545184.0
101989 2021-08-03   Spain      4523310.0
101791 2021-08-02   Spain      4502983.0
101593 2021-08-01   Spain      4447044.0
101395 2021-07-31   Spain      4447044.0

To obtain the daily cases, the cases of one day must be subtracted from those of the previous day.


First I create new columns with the cases from the previous day

In [194]:
df_data = df_data \
      .sort_values(['Country', 'Date']) \
      .assign(ConfirmedPrevious = lambda dataset: dataset.groupby(['Country']).shift(1)["ConfirmedAcum"],
              DeathsPrevious = lambda dataset: dataset.groupby(['Country']).shift(1)["DeathsAcum"],
              RecoveredPrevious = lambda dataset: dataset.groupby(['Country']).shift(1)["RecoveredAcum"],
              ActivePrevious = lambda dataset: dataset.groupby(['Country']).shift(1)["ActiveAcum"],
      ) \
      .fillna({ 'ConfirmedPrevious' : 0, 'DeathsPrevious' : 0, 'RecoveredPrevious' : 0 })

In [195]:
df_data \
      .query("Country == 'Spain'") \
      .sort_values(['Country', 'Date'], ascending = False) \
      .filter(['Date', 'Country','ConfirmedAcum', 'ConfirmedPrevious']) \
      .head(5)

Date Country  ConfirmedAcum  ConfirmedPrevious
102187 2021-08-04   Spain      4545184.0          4523310.0
101989 2021-08-03   Spain      4523310.0          4502983.0
101791 2021-08-02   Spain      4502983.0          4447044.0
101593 2021-08-01   Spain      4447044.0          4447044.0
101395 2021-07-31   Spain      4447044.0          4447044.0

After that I am going to assign the new fields subtracting the previous acum cases to the actual acum cases

In [196]:
df_data = df_data.assign(
      Confirmed = lambda dataset: dataset.ConfirmedAcum -  dataset.ConfirmedPrevious,
      Deaths = lambda dataset: dataset.DeathsAcum - dataset.DeathsPrevious,
      Recovered = lambda dataset: dataset.RecoveredAcum - dataset.RecoveredPrevious,
      Active = lambda dataset: dataset.ActiveAcum - dataset.ActivePrevious
    )

In [200]:
df_data \
      .query("Country == 'Spain'") \
      .sort_values(['Country', 'Date'], ascending = False) \
      .filter(['Date', 'Country','ConfirmedAcum', 'ConfirmedPrevious', 'Confirmed']) \
      .head(5)

Date Country  ConfirmedAcum  ConfirmedPrevious  Confirmed
102187 2021-08-04   Spain      4545184.0          4523310.0    21874.0
101989 2021-08-03   Spain      4523310.0          4502983.0    20327.0
101791 2021-08-02   Spain      4502983.0          4447044.0    55939.0
101593 2021-08-01   Spain      4447044.0          4447044.0        0.0
101395 2021-07-31   Spain      4447044.0          4447044.0        0.0

I no longer need the fields I used to make the calculation so I can drop them

In [201]:
df_data = df_data.drop(columns = ['ConfirmedPrevious', 'DeathsPrevious', 'RecoveredPrevious', 'ActivePrevious'])

Does the data look good?

In [204]:
df_data \
      .query("Country == 'Spain'") \
      .sort_values(['Country', 'Date'], ascending = False) \
      .head(5)

Date Country  ConfirmedAcum  DeathsAcum  RecoveredAcum  \
102187 2021-08-04   Spain      4545184.0     81844.0       150376.0   
101989 2021-08-03   Spain      4523310.0     81773.0       150376.0   
101791 2021-08-02   Spain      4502983.0     81643.0       150376.0   
101593 2021-08-01   Spain      4447044.0     81486.0       150376.0   
101395 2021-07-31   Spain      4447044.0     81486.0       150376.0   

        ActiveAcum  Confirmed  Deaths  Recovered   Active  
102187   4312964.0    21874.0    71.0        0.0  21803.0  
101989   4291161.0    20327.0   130.0        0.0  20197.0  
101791   4270964.0    55939.0   157.0        0.0  55782.0  
101593   4215182.0        0.0     0.0        0.0      0.0  
101395   4215182.0        0.0     0.0        0.0      0.0

## Cases per million inhabitants



We are going to enrich the information with the number of cases per million inhabitants, so we need population data by country.

A small internet search leads me to a page that has population data for 2020:

https://www.worldometers.info/world-population/population-by-country/

It seems that this information is protected to be downloaded automatically so I have no choice but to do it manually and upload the data to a GitHub Repository:

https://github.com/dvillaj/world-population/



I load the data to the Pandas, clean it up and just maintain the field of the population 


In [211]:
df_population = pd.read_excel("https://github.com/dvillaj/world-population/blob/master/data/world-popultation-2020.xlsx?raw=true", sheet_name="Data")

In [212]:
df_population.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 235 entries, 0 to 234
Data columns (total 11 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Country            235 non-null    object 
 1   Population (2020)  235 non-null    int64  
 2   Yearly Change      235 non-null    float64
 3   Net Change         235 non-null    int64  
 4   Density (P/Km²)    235 non-null    float64
 5   Land Area (Km²)    235 non-null    int64  
 6   Migrants (net)     201 non-null    float64
 7   Fertility Rate     201 non-null    float64
 8   Average Age        201 non-null    float64
 9   Urban Pop %        222 non-null    float64
 10  World Share        235 non-null    float64
dtypes: float64(7), int64(3), object(1)
memory usage: 20.3+ KB


In [214]:
df_population = df_population.rename(columns = {
    'Population (2020)' : 'Population',
    'Yearly Change' : 'Yearly_Change',
    'Net Change' : 'Net_Change',
    'Density (P/Km²)' : 'Density',
    'Land Area (Km²)' : 'Land_Area',
    'Migrants (net)' : 'igrants',
    'Fertility Rate' : 'Fertility',
    'Average Age' : 'Mean_Age',
    'Urban Pop %' : 'Urban_Pop',
    'World Share' : 'World_Share'
})

In [217]:
df_population = df_population.filter(['Country', 'Population'])

In [219]:
df_population.head()

Country  Population
0     Afghanistan    38928346
1         Albania     2877797
2         Algeria    43851044
3  American Samoa       55191
4         Andorra       77265

Now I join the population Dataset with the country data to have the population in this dataset

In [220]:
df_data = df_data.merge(df_population, how = 'left', on = 'Country')

In [221]:
df_data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 102220 entries, 0 to 102219
Data columns (total 11 columns):
 #   Column         Non-Null Count   Dtype         
---  ------         --------------   -----         
 0   Date           102220 non-null  datetime64[ns]
 1   Country        102220 non-null  object        
 2   ConfirmedAcum  102220 non-null  float64       
 3   DeathsAcum     102220 non-null  float64       
 4   RecoveredAcum  102220 non-null  float64       
 5   ActiveAcum     102220 non-null  float64       
 6   Confirmed      102220 non-null  float64       
 7   Deaths         102220 non-null  float64       
 8   Recovered      102220 non-null  float64       
 9   Active         101984 non-null  float64       
 10  Population     96828 non-null   float64       
dtypes: datetime64[ns](1), float64(9), object(1)
memory usage: 9.4+ MB


In [222]:
df_data \
      .query("Country == 'Spain'") \
      .sort_values(['Country', 'Date'], ascending = False) \
      .head(5)

Date Country  ConfirmedAcum  DeathsAcum  RecoveredAcum  \
85625 2021-08-04   Spain      4545184.0     81844.0       150376.0   
85624 2021-08-03   Spain      4523310.0     81773.0       150376.0   
85623 2021-08-02   Spain      4502983.0     81643.0       150376.0   
85622 2021-08-01   Spain      4447044.0     81486.0       150376.0   
85621 2021-07-31   Spain      4447044.0     81486.0       150376.0   

       ActiveAcum  Confirmed  Deaths  Recovered   Active  Population  
85625   4312964.0    21874.0    71.0        0.0  21803.0  46754778.0  
85624   4291161.0    20327.0   130.0        0.0  20197.0  46754778.0  
85623   4270964.0    55939.0   157.0        0.0  55782.0  46754778.0  
85622   4215182.0        0.0     0.0        0.0      0.0  46754778.0  
85621   4215182.0        0.0     0.0        0.0      0.0  46754778.0

And finally I calculate the number of cases per million inhabitants

In [233]:
df_data = df_data \
      .assign(ConfirmedAcum_Millon = 
              lambda dataset: (dataset.ConfirmedAcum / dataset.Population * 1000000).round(0))

In [234]:
df_data \
      .query("Country == 'Spain'") \
      .sort_values(['Country', 'Date'], ascending = False) \
      .filter(['Date', 'Country', 'ConfirmedAcum', 'Population', 'ConfirmedAcum_Millon']) \
      .head(5)

Date Country  ConfirmedAcum  Population  ConfirmedAcum_Millon
85625 2021-08-04   Spain      4545184.0  46754778.0               97213.0
85624 2021-08-03   Spain      4523310.0  46754778.0               96745.0
85623 2021-08-02   Spain      4502983.0  46754778.0               96311.0
85622 2021-08-01   Spain      4447044.0  46754778.0               95114.0
85621 2021-07-31   Spain      4447044.0  46754778.0               95114.0

## Last day cases

I'm going to create a dataset of last day's cases. 

The goal is to get a set of rankings that tell me the countries with the most cases

So I need a variable that contains the last date of the dataset

In [235]:
last_day = list(df_data.Date.sort_values(ascending = False))[0]
last_day

Timestamp('2021-08-04 00:00:00')

Now I can filter the data by this date

In [236]:
last_day_data = df_data.query("Date == @last_day")
last_day_data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 198 entries, 527 to 102219
Data columns (total 12 columns):
 #   Column                Non-Null Count  Dtype         
---  ------                --------------  -----         
 0   Date                  198 non-null    datetime64[ns]
 1   Country               198 non-null    object        
 2   ConfirmedAcum         198 non-null    float64       
 3   DeathsAcum            198 non-null    float64       
 4   RecoveredAcum         198 non-null    float64       
 5   ActiveAcum            198 non-null    float64       
 6   Confirmed             198 non-null    float64       
 7   Deaths                198 non-null    float64       
 8   Recovered             198 non-null    float64       
 9   Active                198 non-null    float64       
 10  Population            188 non-null    float64       
 11  ConfirmedAcum_Millon  188 non-null    float64       
dtypes: datetime64[ns](1), float64(10), object(1)
memory usage: 20.1+ KB


Which countries have the most confirmed cases?

In [238]:
last_day_data \
  .sort_values('ConfirmedAcum', ascending = False) \
  .filter(['Country', 'ConfirmedAcum']) \
  .reset_index(drop = True) \
  .head(10)

Country  ConfirmedAcum
0   United States     35453552.0
1           India     31812114.0
2          Brazil     20034407.0
3          Russia      6274006.0
4          France      6272466.0
5  United Kingdom      5980830.0
6          Turkey      5822487.0
7       Argentina      4975616.0
8        Colombia      4815063.0
9           Spain      4545184.0

Which countries have the most active cases?

In [239]:
last_day_data \
  .sort_values('ActiveAcum', ascending = False) \
  .filter(['Country', 'ActiveAcum']) \
  .reset_index(drop = True) \
  .head(10)

Country  ActiveAcum
0  United Kingdom   5825839.0
1          France   5745110.0
2           Spain   4312964.0
3     Netherlands   1858347.0
4          Brazil   1703235.0
5         Belgium   1107676.0
6          Sweden   1088172.0
7          Serbia    716389.0
8        Thailand    640009.0
9       Indonesia    524011.0

Which countries have the most active cases per million inhabitants?

In [241]:
last_day_data \
  .sort_values('ConfirmedAcum_Millon', ascending = False) \
  .filter(['Country', 'ConfirmedAcum_Millon']) \
  .reset_index(drop = True) \
  .head(10)

Country  ConfirmedAcum_Millon
0                   Andorra              191510.0
1                Seychelles              188577.0
2                Montenegro              163454.0
3                   Bahrain              158451.0
4  Czech Republic (Czechia)              156334.0
5                San Marino              152604.0
6                  Maldives              144105.0
7                  Slovakia              142265.0
8                  Slovenia              124883.0
9                Luxembourg              118445.0

What is the total number of cases?

In [250]:
last_day_data.agg(
    TotalConfirmed = ("ConfirmedAcum", "sum"),
    TotalRecovered = ("RecoveredAcum", "sum"),
    TotalDeaths = ("DeathsAcum", "sum"),
) \
.style.format("{:,.0f}")

## Exporting to Excel the clean dataset

Finally I will create an Excel file with the information per country once it is clean and in perfect condition to apply some machine learning algorithms ...

In [251]:
df_data.to_excel("All_data.xlsx", index = False)